mlir_graphblas/src/lib/GraphBLAS/GraphBLASLowerPass.cpp

```diff
@@ -468,58 +469,109 @@ class LowerCastRewrite : public OpRewritePattern<graphblas::CastOp> {
+        op->getResultTypes().front().dyn_cast<RankedTensorType>();
+
     sparse_tensor::SparseTensorEncodingAttr sparseEncoding =
         sparse_tensor::getSparseTensorEncoding(inputType);
     unsigned ptrBitWidth = sparseEncoding.getPointerBitWidth();
     unsigned idxBitWidth = sparseEncoding.getIndexBitWidth();
     Type inputValueType = inputType.getElementType();
-    RankedTensorType outputType =
-        op->getResultTypes().front().dyn_cast<RankedTensorType>();
+    llvm::ArrayRef<int64_t> inputShape = inputType.getShape();
 
     bool inputTypeIsCSR = typeIsCSR(inputType);
-    bool outputTypeIsCSR = typeIsCSR(outputType);
 
     RankedTensorType flippedInputType =
         getSingleCompressedMatrixType(context, inputShape, inputTypeIsCSR,
                                       inputValueType, ptrBitWidth, idxBitWidth);
 
-    // Add a graphblas.convert_layout op if the input and output compression
-    // types are the same
-    if (inputTypeIsCSR == outputTypeIsCSR) {
-      // TODO consider separating this out into its own rewrite pattern
-      Value flippedInput = rewriter.create<graphblas::ConvertLayoutOp>(
-          loc, flippedInputType, inputTensor);
-      Value transposed = rewriter.create<graphblas::TransposeOp>(
-          loc, outputType, flippedInput);
+    Value flippedInput = rewriter.create<graphblas::ConvertLayoutOp>(
+        loc, flippedInputType, inputTensor);
+    Value transposed =
+        rewriter.create<graphblas::TransposeOp>(loc, outputType, flippedInput);
+
+    rewriter.replaceOp(op, transposed);
+  };
+};
 
-      rewriter.replaceOp(op, transposed);
+class LowerTransposeRewrite : public OpRewritePattern<graphblas::TransposeOp> {
+public:
+  using OpRewritePattern<graphblas::TransposeOp>::OpRewritePattern;
 
+  LogicalResult match(graphblas::TransposeOp op) const override {
+    if (LowerTransposeLayoutDWIMRewrite::needsDWIM(op))
+      return failure();
+    else
       return success();
-    }
+  };
+
+  void rewrite(graphblas::TransposeOp op,
+               PatternRewriter &rewriter) const override {
+
+    MLIRContext *context = op.getContext();
+    ModuleOp module = op->getParentOfType<ModuleOp>();
+    Location loc = op->getLoc();
+
+    Value inputTensor = op.input();
+    RankedTensorType inputType =
+        inputTensor.getType().dyn_cast<RankedTensorType>();
+    llvm::ArrayRef<int64_t> inputShape = inputType.getShape();
+    sparse_tensor::SparseTensorEncodingAttr sparseEncoding =
+        sparse_tensor::getSparseTensorEncoding(inputType);
+    unsigned ptrBitWidth = sparseEncoding.getPointerBitWidth();
+    unsigned idxBitWidth = sparseEncoding.getIndexBitWidth();
+    Type inputValueType = inputType.getElementType();
+
+    bool inputTypeIsCSR = typeIsCSR(inputType);
+
+    RankedTensorType flippedInputType =
+        getSingleCompressedMatrixType(context, inputShape, inputTypeIsCSR,
+                                      inputValueType, ptrBitWidth, idxBitWidth);
```

Comment:
I don't think we need to manually build this. Calling `op.getResult().getType().cast<RankedTensorType>()` should give us the flipped type. Otherwise `match` would have failed.

Reply:
That makes sense. I'll implement this suggestion.

Reply:
https://github.com/metagraph-dev/mlir-graphblas/commit/d5e602fea2f3e2514222e7b65c8dbda36922fbd6 implements this suggestion.